indexing-service/src/main/java/org/apache/druid/indexing/common/task/RangePartitionCachingLocalSegmentAllocator.java

```diff
@@ -0,0 +1,195 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.indexing.common.task;
+
+import com.google.common.collect.Maps;
+import org.apache.druid.data.input.InputRow;
+import org.apache.druid.indexing.common.TaskToolbox;
+import org.apache.druid.segment.realtime.appenderator.SegmentIdWithShardSpec;
+import org.apache.druid.timeline.partition.SingleDimensionShardSpec;
+import org.joda.time.Interval;
+
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+/**
+ * Allocates all necessary range-partitioned segments locally at the beginning and reuses them.
+ *
+ * @see CachingLocalSegmentAllocator
+ */
+public class RangePartitionCachingLocalSegmentAllocator implements IndexTaskSegmentAllocator
+{
+  private final String dataSource;
+  private final String partitionDimension;
+  private final Map<Interval, String[]> intervalsToPartitions;
+  private final IndexTaskSegmentAllocator delegate;
+
+  public RangePartitionCachingLocalSegmentAllocator(
+      TaskToolbox toolbox,
+      String taskId,
+      String supervisorTaskId,
+      String dataSource,
+      String partitionDimension,
+      Map<Interval, String[]> intervalsToPartitions
+  ) throws IOException
+  {
+    this.dataSource = dataSource;
+    this.partitionDimension = partitionDimension;
+    this.intervalsToPartitions = intervalsToPartitions;
+
+    this.delegate = new CachingLocalSegmentAllocator(
+        toolbox,
+        taskId,
+        supervisorTaskId,
+        this::getIntervalToSegmentIds
+    );
+  }
+
+  private Map<Interval, List<SegmentIdWithShardSpec>> getIntervalToSegmentIds(Function<Interval, String> versionFinder)
+  {
+    Map<Interval, List<SegmentIdWithShardSpec>> intervalToSegmentIds =
+        Maps.newHashMapWithExpectedSize(intervalsToPartitions.size());
+
+    intervalsToPartitions.forEach(
+        (interval, partitions) ->
+            intervalToSegmentIds.put(
+                interval,
+                translatePartitions(interval, partitions, versionFinder)
+            )
+    );
+
+    return intervalToSegmentIds;
+  }
+
+  private List<SegmentIdWithShardSpec> translatePartitions(
+      Interval interval,
+      String[] partitions,
+      Function<Interval, String> versionFinder
+  )
+  {
+    if (partitions.length == 0) {
+      return Collections.emptyList();
+    }
+
+    String[] uniquePartitions = Arrays.stream(partitions).distinct().toArray(String[]::new);
```

Comment:
Currently, `StringSketch.getEventPartitionsByCount()` can return duplicate values since it's a wrapper for `ItemsSketch.getQuantiles()` (i.e., if the distribution has many duplicates, adjacent quantiles may have the same value).

I'll also fix the typo in "getEventPartitionsByCount".